libgalois/include/galois/graphs/FileGraph.h

```diff
@@ -700,82 +699,58 @@ class FileGraphWriter : public FileGraph {
   void setNumEdges(size_t n) { numEdges = n; }
   //! Set the size of the edge data to write to n
   //! @param n size of edge data to write
-  void setSizeofEdgeData(size_t n) { sizeofEdgeData = n; }
+  void setSizeofEdgeData(size_t n) { sizeofEdge = n; }
 
   //! Marks the transition to next phase of parsing: counting the degree of
   //! nodes
-  void phase1() { outIdx.resize(numNodes); }
+  void phase1();
 
   //! Increments degree of id by delta
-  void incrementDegree(size_t id, int delta = 1) {
+  void incrementDegree(size_t id, uint64_t delta = 1) {
     assert(id < numNodes);
     outIdx[id] += delta;
   }
 
   //! Marks the transition to next phase of parsing, adding edges
-  void phase2() {
-    if (numNodes == 0)
-      return;
-
-    // Turn counts into partial sums
-    auto prev = outIdx.begin();
-    for (auto ii = outIdx.begin() + 1, ei = outIdx.end(); ii != ei;
-         ++ii, ++prev) {
-      *ii += *prev;
-    }
-    assert(outIdx[numNodes - 1] == numEdges);
-
-    if (numNodes <= std::numeric_limits<uint32_t>::max()) {
-      // version 1
-      starts.resize(numNodes);
-      outs.resize(numEdges);
-    } else {
-      // version 2
-      starts64.resize(numNodes);
-      outs64.resize(numEdges);
-    }
-  }
+  void phase2();
 
   //! Adds a neighbor between src and dst
   size_t addNeighbor(size_t src, size_t dst) {
     size_t base = src ? outIdx[src - 1] : 0;
+    size_t idx  = base + starts[src]++;
+    assert(idx < outIdx[src]);
+
+    if (numNodes <= std::numeric_limits<uint32_t>::max())
+      reinterpret_cast<uint32_t*>(outs)[idx] = dst; // version 1
+    else
+      reinterpret_cast<uint64_t*>(outs)[idx] = dst; // version 2
+    return idx;
+  }
 
-    if (numNodes <= std::numeric_limits<uint32_t>::max()) {
-      // version 1
-      size_t idx = base + starts[src]++;
-      assert(idx < outIdx[src]);
-      outs[idx] = dst;
-      return idx;
-    } else {
-      // version 2
-      size_t idx = base + (starts64)[src]++;
-      assert(idx < outIdx[src]);
-      outs64[idx] = dst;
-      return idx;
-    }
+  //! Adds a neighbor between src and dst w/ corresponding data
+  template <typename T>
+  size_t addNeighbor(
```

Comment:
You have both a templated and non-templated function with the same name. Why does this not cause a compiler error? Why not handle it by a special case of edge data being void type?

Reply:
A better design might be to have a separate addEdgeData function which takes an index (that is returned by addNeighbor) and data.

Reply:
Good question :)!

Remember the reference link you showed me about the virtual stuff? I learned something about [template functions](http://www.gotw.ca/publications/mill17.htm) there as well.

If I understand the material currently and what it suggests also makes sense to you, what I learned is that *non-template overload has the first priority than the templates* - that's why this does not cause a compiler error.

The deal I'm handling: For void, addNeighbor only requires *src* and *dst* and does only adding the edge; for non-void, addNeighbor could require *src*, *dst*, and *data*, and does not only adding the edge but also adding the edge data. The non-void case should also use the addNeighbor for void case without adding the edge data when adding the edge, but it does not make sense for void case to have addNeighbor with *src*, *dst*, and *data* (therefore I did not provide this specialization).

In my design, non-template addNeighbor could be used for both void and non-void cases and it's always called prior to the template one. The template one only supports non-void cases and compiler errors for void cases.

A separate addEdgeData may not prevent users from calling it for void cases, though assertion could be made at the beginning of its definition.

Reply:
Thanks for the detailed explanation. This choice/design is fine, although I like the previous design of a separate addEdgeData function which takes an index (that is returned by addNeighbor) and data.

Reply:
I'll add the separate addEdgeData function as an alternative but still keep my choice :) Thanks!

Reply:
No, don't do that! Go with one of them.

Typically, there should NOT be two ways to do the same thing.

Reply:
In that case, I'd like to insist on my choice unless a more convincing objection :)

Reply:
Yes, that's fine.